### DISCUSSION TITLE
ESP32-C3 + WS2816 issues

### DISCUSSION THREAD

**Discussion by mcer12** [Makuna/NeoPixelBus#708] [Q&A]
First off, I'm sorry if this is the wrong place to discuss this.

In the Wiki -> ESP32 NeoMethods there are some limitations mentioned about C3 module but it should generally support I2S, RMT and bitbang but I wasn't able to use neither of those. Bitbang is working but unusable (glitches), RMT and I2S don't crash but don't do anythig. The only thing that works is the default NeoWs2816Method. Searching through issues, I2S seems to be unsupported but RMT should work... which it doesn't, at least for me. Maybe the documentation is correct but there's an issue with WS2816 implementation in particular? Tested on NeoPixelTest example.

**Comment by Makuna** [comment]
Its best to start these is the discussion area first. Once its know its a bug in the code, an issue could be created.

ESP32-S3 and ESP32-C3 do not support i2s yet.  There is an issue tracking that already.

https://github.com/Makuna/NeoPixelBus/wiki/ESP32-NeoMethods does not correctly state this (it may by the time you read this though).

It helps if you provide the snippit of code that shows how you define your neopixelbus for each case.

NeoWs2816Method maps to RMT channel 1 on S2/C3/S3, so it if works then RMT works.  It is the same as NeoEsp32Rmt1Ws2812xMethod.

**Reply by mcer12** [reply]
Sorry I'm getting back to this after so long, I've been moving around alot and didn't have time for anything... I got back to WS2816 and investigated further. The setup is 225 WS2816C leds in series using good quality 5V PSU. RMT method works but colors with values above about 160 start glitching (only roughly second half of the strip glitches, first half of the strip never glitches). Since this is fine with some values and not with others, I don't think it's a faulty led. 

ie this works fine:
Rgb48Color test(10, 10, 10);
This glitches:
Rgb48Color test(5000, 5000, 5000);

To make sure it's not some issue with ESP32C3 I replaced the module with ESP8266, swith exactly the samebehavior. Setting led count to some higher value changes the pattern of the glitches. This leads me to think there's some issue with timing. According to datasheet it is slightly different from WS2812 but in the code WS2816 and WS2812x seem interchangeable.

**Reply by mcer12** [reply]
Here's the code for ESP8266 (using inverted method with a single transistor level shifter):

`#include <NeoPixelBus.h>

const uint16_t PixelCount = 225; // this example assumes 4 pixels, making it smaller will cause a failure
const uint8_t PixelPin = 3;  // make sure to set this to the correct pin, ignored for Esp8266

NeoPixelBus<NeoGrb48Feature, NeoWs2816InvertedMethod> strip(PixelCount, PixelPin);

Rgb48Color white(1159, 1159, 1159);
Rgb48Color black(0, 0, 0);

void setup()
{

  // this resets all the neopixels to an off state
  strip.Begin();
  strip.Show();

}


void loop()
{
  for (int i = 0; i < PixelCount; i++) {
    strip.SetPixelColor(i, white);
  }
  strip.Show();

  delay(10);

  for (int i = 0; i < PixelCount; i++) {
    strip.SetPixelColor(i, black);
  }
  strip.Show();
  delay(10);

}`

**Reply by Makuna** [reply]
Exactly how are you connecting the power?  For these longer strips you need to connect power to both ends, and sometimes even in the center.  The strips just don't use thick enough traces.  Of course, make sure the GND of the ESP connects to the GND of the strip also.

The ESP8266 uses i2S with very limited ability to change the signal pulse form.  The ESP32 uses RMT with a pretty accurate signal pulse form.  If they are acting the same there is a higher chance its external issue; but I will take a glance.

**Reply by mcer12** [reply]
It's a custom board, here's how it looks. I maybe wasn't clear that the "flicker" only occurs with update as in random data is passed to the second half of the strip and stays that way until next update. Sometimes the update is successful without any glitches. I can make a video if that helps.
![Snímek obrazovky 2023-09-03 123951](https://github.com/Makuna/NeoPixelBus/assets/33605121/08d99984-f47f-417c-8730-19f5cf513992)

**Reply by Makuna** [reply]
BTW, i2S is not supported on C3 at all right now.  So, I am unsure how you tried it as it should not have compiled if you used an i2s specific method name.  The unspecific names like NeoWs2816Method uses RMT on C3 and S3.

You are using the latest version of the library, correct?  

I believe I see the wider positive power bus, but where is the equally sized GND?  Is it the rest of the surface (ground plane)?  Backside?  How does it get to the third row from the bottom as the signal wires isolate that plane?  Again, Vias to the other side?

When you state it works fine with low levels of color `Rgb48Color(10,10,10)` and not with higher levels like `Rgb48Color(5000,5000,5000)` this indicates to me that it's a power distribution issue.  Either too small of a power supply or poor GND/+ wire sizing.  What about `Rgb48Color(512,512,512)`?  

225 pixels => 60ma (@5v) x 225 = 13.5A.  that's over 13 AMPs of power.  This is max at full bright, so a worst-case scenario; but a good engineering practice is to provide more than you need by a small margin.

**Reply by mcer12** [reply]
@Makuna whole bottom side of the board is a GND plane. Please note that I'm using WS2816C which is a low current version (12mA per led max accoring to datasheet, that means about 2.7A max), power supply is 5A rated. There's also a 3A fuse onboard, which doesn't blow :)

**Reply by mcer12** [reply]
I am using latest ESP8266/ESP32 arduino packages as well as latest library. Yes I understand I2S isn't working for ESP32C3, I was using RMT method there.

**Reply by mcer12** [reply]
@Makuna I replaced and tested few leds that seemed they could be causing the glitching and it's solved! Full brightness, no glitches! So either they were faulty or there was some hidden cold joint (i did solder these by hand). What I don't understand is why it worked fine at low brightness but who cares. I'll keep it running for few hours, see if it's really not doing anything weird anymore and close the discussion. Thank you very much for your help and for insisting on it being a HW problem.

**Comment by Makuna** [comment]
Since I have a test case (you), would you be willing to try a small change for RMT to better fit the spec?  I will put a formal branch, but you could just change NeoEsp32RmtSpeedWs2812x class in the NeoEsp32RmtMethod.h to test it.

Change the NeoEsp32RmtSpeedWs2812x class in the NeoEsp32RmtMethod.h to this
```
class NeoEsp32RmtSpeedWs2812x : public NeoEsp32RmtSpeedBase
{
public:
    const static DRAM_ATTR uint32_t RmtBit0 = Item32Val(260, 1000);
    const static DRAM_ATTR uint32_t RmtBit1 = Item32Val(660, 600);
    const static DRAM_ATTR uint16_t RmtDurationReset = FromNs(280000); // 280us

    static void IRAM_ATTR Translate(const void* src,
        rmt_item32_t* dest,
        size_t src_size,
        size_t wanted_num,
        size_t* translated_size,
        size_t* item_num);
};
```

Then give it a try and see if it still works.

**Reply by mcer12** [reply]
I'll try that but I'll have to first solder the C3 module back :) If you're wondering, ESP32-C3-WROOM-02 and ESP32-WROOM-02 are somewhat pin compatible. I'd leave the reset at 300us because the datasheet says >280us.

**Reply by Makuna** [reply]
The reset time in the structure is a minimum, at best it usually takes another 20us and at worst another 80us.  All depends on byte boundaries from count and size of pixels and rounding up to size boundaries DMA requires.  And a little code overhead.

AND note, you can use RMT for any ESP32 variant, it's the i2S that is limited to ESP32, C2, and S2.
